multicluster/controllers/multicluster/gateway_controller.go

```diff
@@ -0,0 +1,328 @@
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=gateways/finalizers,verbs=update
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterinfoimports,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterinfoimports/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=multicluster.crd.antrea.io,resources=clusterinfoimports/finalizers,verbs=update
+
+func (r *GatewayReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	klog.InfoS("Reconciling Gateway", "gateway", req.NamespacedName)
+	requeue, remoteCluster, err := r.checkRemoteManager()
+	if requeue {
+		return ctrl.Result{Requeue: true}, nil
+	}
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	r.leaderNamespace = remoteCluster.GetNamespace()
+	err = r.getServiceCIDR(ctx)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	resExportName := r.localClusterID + "-clusterinfo"
+	resExportNamespaced := types.NamespacedName{
+		Name:      resExportName,
+		Namespace: r.leaderNamespace,
+	}
+	resExport := &mcsv1alpha1.ResourceExport{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      resExportName,
+			Namespace: r.leaderNamespace,
+		},
+	}
+
+	createOrUpdate := func(gwIP string) error {
+		existingResExport := &mcsv1alpha1.ResourceExport{}
+		if err := remoteCluster.Get(ctx, resExportNamespaced, existingResExport); err != nil {
+			if !apierrors.IsNotFound(err) {
+				return err
+			}
+			if err = r.createorUpdateResourceExport(ctx, req, remoteCluster, existingResExport, gwIP, true); err != nil {
+				return err
+			}
+			return nil
+		}
+		if err = r.createorUpdateResourceExport(ctx, req, remoteCluster, existingResExport, gwIP, false); err != nil {
+			return err
+		}
+		return nil
+	}
+
+	gwn := &mcsv1alpha1.Gateway{}
+	if err := r.Client.Get(ctx, req.NamespacedName, gwn); err != nil {
+		if apierrors.IsNotFound(err) {
+			gwNodeInfo, err := r.getLastCreatedGateway()
+			if err != nil {
+				klog.ErrorS(err, "Failed to get Gateways")
+				return ctrl.Result{}, err
+			}
+			if gwNodeInfo == nil {
+				// When the last Gateway is deleted, we will remove the ClusterInfo kind of ResourceExport
+				if err := remoteCluster.Delete(ctx, resExport, &client.DeleteOptions{}); err != nil {
+					return ctrl.Result{}, client.IgnoreNotFound(err)
+				}
+				return ctrl.Result{}, nil
+			}
+			// When there are still Gateways exist, we should create or update existing ResourceExport
+			// with the latest Gateway in remaining Gateways.
+			if err := createOrUpdate(gwNodeInfo.GatewayIP); err != nil {
+				return ctrl.Result{}, err
+			}
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if err := createOrUpdate(gwn.GatewayIP); err != nil {
+		return ctrl.Result{}, err
+	}
+	return ctrl.Result{}, nil
+}
+
+func (r *GatewayReconciler) checkRemoteManager() (bool, commonarea.RemoteCommonArea, error) {
+	defer r.mutex.Unlock()
+	r.mutex.Lock()
```

Comment:
If we have multiple routines can read remoteCommonAreaManager in parallel (?), we should consider RW lock.

Reply:
I do not see you lock when remoteCommonAreaManager is set?

Reply:
remoteCommonAreaManager is initialized in member_clusterset_controller and it's already protected by mutex.

Reply:
done, change mutex to RWMutex.